arches/app/models/graph.py

```diff
@@ -1606,6 +1607,35 @@ def get_nodegroups(self, force_recalculation=False):
                     pass
             return list(nodegroups)
 
+    @contextmanager
+    @transaction.atomic
+    def preserve_staging_records(self):
+        nodegroups = self.get_nodegroups(force_recalculation=True)
+        error_query = models.LoadErrors.objects.filter(nodegroup__in=nodegroups)
+        staging_query = models.LoadStaging.objects.filter(nodegroup__in=nodegroups)
+        error_objs = set(error_query)
+        staging_objs = set(staging_query)
+        error_query.update(nodegroup=None)
+        staging_query.update(nodegroup=None)
+
+        try:
+            yield
+        finally:
+            # Restore the nodegroup references that still exist.
+            all_nodegroup_ids = models.NodeGroup.objects.values_list("pk", flat=True)
+            valid_errors = {
+                obj for obj in error_objs if obj.nodegroup_id in all_nodegroup_ids
+            }
+            valid_stagings = {
+                obj for obj in staging_objs if obj.nodegroup_id in all_nodegroup_ids
+            }
+            invalid_errors = error_objs - valid_errors
+            invalid_stagings = staging_objs - valid_stagings
+            models.LoadErrors.objects.bulk_update(valid_errors, fields=["nodegroup"])
+            models.LoadStaging.objects.bulk_update(valid_stagings, fields=["nodegroup"])
+            models.LoadErrors.objects.filter(pk__in=invalid_errors).delete()
+            models.LoadStaging.objects.filter(pk__in=invalid_stagings).delete()
```

Comment:
Can also just delete these two lines if we'd prefer leaving these records with null nodegroups. cc/ @chiatt

Reply:
Letting these stay null rather than delete after chatting w/ Cyrus.